File: app/src/main/java/com/android/learn/mpresenter/KnowledgeChildPresenter.java
package com.android.learn.mpresenter;

import com.android.base.mmodel.BaseData;
import com.android.base.mmodel.BaseResponse;
import com.android.base.mmodel.FeedArticleListData;
import com.android.base.mmodel.FeedArticleListData.FeedArticleData;
import com.android.base.mpresenter.BasePresenter;
import com.android.base.thirdframe.rxjava.BaseObserver;
import com.android.base.utils.ResponseStatusUtil;
import com.android.learn.mcontract.KnowledgeChildContract;

import io.reactivex.Observable;


public class KnowledgeChildPresenter extends BasePresenter<KnowledgeChildContract.View> implements KnowledgeChildContract.Presenter {
    private int mCurrentPage = 0;

    @Override
    public void getKnowledgeArticleList(int num, int cid) {
        Observable observable = mRestService.getKnowledgeArticleList(num, cid);
        addSubscribe(observable, new BaseObserver<BaseResponse<FeedArticleListData>>() {
            @Override
            public void onNext(BaseResponse<FeedArticleListData> feedArticleListData) {
                mView.showArticleList(feedArticleListData.getData(), false);
            }
        });
    }

    @Override
    public void onRefreshMore(int cid) {
        Observable observable = mRestService.getKnowledgeArticleList(-1, cid);
        addSubscribe(observable, new BaseObserver<BaseResponse<FeedArticleListData>>() {
            @Override
            public void onNext(BaseResponse<FeedArticleListData> feedArticleListData) {
                mView.showArticleList(feedArticleListData.getData(), true);
            }
        });


    }

    @Override
    public void onLoadMore(int cid) {
        ++mCurrentPage;
        Observable observable = mRestService.getKnowledgeArticleList(mCurrentPage, cid);
        addSubscribe(observable, new BaseObserver<BaseResponse<FeedArticleListData>>() {
            @Override
            public void onNext(BaseResponse<FeedArticleListData> feedArticleListData) {
                mView.showArticleList(feedArticleListData.getData(), false);
            }
        });
    }

    @Override
    public void addCollectArticle(final int position, final FeedArticleData feedArticleData) {
        Observable observable = mRestService.addCollectArticle(feedArticleData.getId());
        addSubscribe(observable, new BaseObserver<BaseData>() {

            @Override
            public void onNext(BaseData data) {
                if (data.errorCode == BaseData.SUCCESS) {
                    feedArticleData.setCollect(true);
                    mView.showCollectArticleData(position, feedArticleData);
                } else ResponseStatusUtil.handleResponseStatus(data);
            }

        });

    }

    @Override
    public void cancelCollectArticle(final int position, final FeedArticleData feedArticleData) {
        Observable observable = mRestService.cancelCollectArticle(feedArticleData.getId(), -1);
        addSubscribe(observable, new BaseObserver<BaseData>() {

            @Override
            public void onNext(BaseData data) {
                if (data.errorCode == BaseData.SUCCESS) {
                    feedArticleData.setCollect(false);
                    mView.showCancelCollectArticleData(position, feedArticleData);
                } else ResponseStatusUtil.handleResponseStatus(data);
            }

        });
    }


}

File: app/src/main/java/com/android/learn/mpresenter/KnowledgeChildPresenter.kt
package com.android.learn.mpresenter

import com.android.learn.base.mmodel.BaseData
import com.android.learn.base.mmodel.BaseResponse
import com.android.learn.base.mmodel.FeedArticleListData
import com.android.learn.base.mmodel.FeedArticleListData.FeedArticleData
import com.android.learn.base.mpresenter.BasePresenter
import com.android.learn.base.thirdframe.rxjava.BaseObserver
import com.android.learn.base.utils.ResponseStatusUtil
import com.android.learn.mcontract.KnowledgeChildContract

import io.reactivex.Observable


class KnowledgeChildPresenter : BasePresenter<KnowledgeChildContract.View>(), KnowledgeChildContract.Presenter {
    private var mCurrentPage = 0

    override fun getKnowledgeArticleList(num: Int, cid: Int) {
        val observable = mRestService.getKnowledgeArticleList(num, cid)
        addSubscribe(observable, object : BaseObserver<BaseResponse<FeedArticleListData>>(true) {
            override fun onNext(feedArticleListData: BaseResponse<FeedArticleListData>) {
                mView!!.showArticleList(feedArticleListData.data, false)
            }
        })
    }

    override fun onRefreshMore(cid: Int) {
        val observable = mRestService.getKnowledgeArticleList(-1, cid)
        addSubscribe(observable, object : BaseObserver<BaseResponse<FeedArticleListData>>(false) {
            override fun onNext(feedArticleListData: BaseResponse<FeedArticleListData>) {
                mView!!.showArticleList(feedArticleListData.data, true)
            }
        })


    }

    override fun onLoadMore(cid: Int) {
        ++mCurrentPage
        val observable = mRestService.getKnowledgeArticleList(mCurrentPage, cid)
        addSubscribe(observable, object : BaseObserver<BaseResponse<FeedArticleListData>>(false) {
            override fun onNext(feedArticleListData: BaseResponse<FeedArticleListData>) {
                mView!!.showArticleList(feedArticleListData.data, false)
            }
        })
    }

    override fun addCollectArticle(position: Int, feedArticleData: FeedArticleData) {
        val observable = mRestService.addCollectArticle(feedArticleData.id)
        addSubscribe(observable, object : BaseObserver<BaseData>(true) {

            override fun onNext(data: BaseData) {
                if (data.errorCode == BaseData.SUCCESS) {
                    feedArticleData.isCollect = true
                    mView!!.showCollectArticleData(position, feedArticleData)
                } else
                    ResponseStatusUtil.handleResponseStatus(data)
            }

        })

    }

    override fun cancelCollectArticle(position: Int, feedArticleData: FeedArticleData) {
        val observable = mRestService.cancelCollectArticle(feedArticleData.id, -1)
        addSubscribe(observable, object : BaseObserver<BaseData>(true) {

            override fun onNext(data: BaseData) {
                if (data.errorCode == BaseData.SUCCESS) {
                    feedArticleData.isCollect = false
                    mView!!.showCancelCollectArticleData(position, feedArticleData)
                } else
                    ResponseStatusUtil.handleResponseStatus(data)
            }

        })
    }


}
